File: app/src/main/java/sfllhkhan95/doodle/view/PaintView.java
package sfllhkhan95.doodle.view;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.PointF;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.view.View;

import sfllhkhan95.doodle.core.CanvasActionListener;
import sfllhkhan95.doodle.core.PaintBrush;
import sfllhkhan95.doodle.core.PaintCanvas;
import sfllhkhan95.doodle.core.Shapes;
import sfllhkhan95.doodle.shapes.ColorPicker;
import sfllhkhan95.doodle.shapes.Eraser;
import sfllhkhan95.doodle.shapes.Shape;
import sfllhkhan95.doodle.shapes.ShapeFactory;
import sfllhkhan95.doodle.utils.OnColorPickedListener;


/**
 * @author alichishti
 */
public class PaintView extends View {

    private static final float TOUCH_TOLERANCE = 5;
    private final Shapes shapes = new Shapes();
    private final PaintBrush mBrush;

    private PaintCanvas mCanvas;
    private PointF iTouch = new PointF();
    private PointF fTouch = new PointF();
    private Class<? extends Shape> shapeType = null;

    private CanvasActionListener canvasActionListener;
    private OnColorPickedListener onColorPickedListener;

    public PaintView(Context context) {
        super(context);
        mBrush = new PaintBrush(context);
    }

    public PaintView(Context context, AttributeSet attrs) {
        super(context, attrs);
        mBrush = new PaintBrush(context);
    }

    public PaintBrush getBrush() {
        return mBrush;
    }

    public PaintCanvas getCanvas() {
        return mCanvas;
    }

    public void setCanvas(PaintCanvas canvas) {
        mCanvas = canvas;
    }

    public void setShapeType(Class<? extends Shape> shapeType) {
        this.shapeType = shapeType;
    }

    @Override
    protected void onDraw(Canvas canvas) {
        canvas.save();

        mCanvas.drawBackground();
        mCanvas.drawShapes(shapes);

        canvas.drawBitmap(mCanvas.getBitmap(), 0, 0, null);
        canvas.restore();
    }

    @Override
    public boolean performClick() {
        return super.performClick();
    }

    @Override
    public boolean onTouchEvent(MotionEvent event) {
        if (shapeType == null) return false;
        PointF touchAt = new PointF(event.getX(), event.getY());

        switch (event.getAction()) {
            case MotionEvent.ACTION_DOWN:
                touchStart(touchAt);
                performClick();
                invalidate();
                return true;
            case MotionEvent.ACTION_MOVE:
                touchMove(touchAt);
                performClick();
                invalidate();
                return true;
            case MotionEvent.ACTION_UP:
                touchUp(touchAt);
                performClick();
                invalidate();
                return true;
        }

        return false;
    }

    private void touchStart(PointF touchAt) {
        if (this.shapeType == ColorPicker.class) {
            int color = mCanvas.getColor(touchAt);
            if (this.onColorPickedListener != null) {
                this.onColorPickedListener.onColorPicked(color);
            }
            return;
        }

        iTouch.set(touchAt);
        fTouch = null;
    }

    private void touchMove(PointF touchAt) {
        if (this.shapeType == ColorPicker.class) {
            int color = mCanvas.getColor(touchAt);
            if (this.onColorPickedListener != null) {
                this.onColorPickedListener.onColorPicked(color);
            }
            return;
        }


        if (fTouch == null) {
            fTouch = new PointF();

            Shape shape = ShapeFactory.get(shapeType, mBrush);
            shape.moveTo(iTouch.x, iTouch.y);
            if (shape.getClass().equals(Eraser.class)) {
                ((Eraser) shape).initEraser(mCanvas);
            }
            shapes.add(shape);
        }

        float dx = Math.abs(touchAt.x - fTouch.x);
        float dy = Math.abs(touchAt.y - fTouch.y);
        if (dx < TOUCH_TOLERANCE && dy < TOUCH_TOLERANCE) {
            return;
        }

        fTouch.set(touchAt);
        Shape currentShape = shapes.getCurrent();
        if (currentShape != null) {
            currentShape.draw(iTouch, fTouch);
        }
    }

    private void touchUp(PointF touchAt) {
        if (this.shapeType == ColorPicker.class) {
            int color = mCanvas.getColor(touchAt);
            if (this.onColorPickedListener != null) {
                this.onColorPickedListener.onColorPicked(color);
            }
            return;
        }

        if (iTouch == null) return;
        if (fTouch == null) {
            fTouch = touchAt;
            Shape shape = ShapeFactory.get(shapeType, mBrush);
            shape.moveTo(touchAt.x, touchAt.y);
            if (shape.getClass().equals(Eraser.class)) {
                ((Eraser) shape).initEraser(mCanvas);
            }
            shapes.add(shape);
        }

        Shape currentShape = shapes.getCurrent();
        if (currentShape != null) {
            currentShape.draw(iTouch, fTouch);
            if (canvasActionListener != null) {
                canvasActionListener.onDrawPath();
            }
        }
    }

    public void clear() {
        shapes.clear();
        invalidate();

        if (canvasActionListener != null) {
            canvasActionListener.onRevert();
        }
    }

    public void redo() {
        boolean canRedo = shapes.redo();
        invalidate();

        if (canvasActionListener != null) {
            canvasActionListener.onRedo(canRedo);
        }
    }

    public void undo() {
        boolean canUndo = shapes.undo();
        invalidate();

        if (canvasActionListener != null) {
            canvasActionListener.onUndo(canUndo);
        }
    }

    public void save() {
        mCanvas.saveProject();
    }

    public void setCanvasActionListener(CanvasActionListener canvasActionListener) {
        this.canvasActionListener = canvasActionListener;
    }

    public void setOnColorPickedListener(OnColorPickedListener onColorPickedListener) {
        this.onColorPickedListener = onColorPickedListener;
    }

    public boolean isModified() {
        return shapes.getPointer() > 0;
    }

    public boolean canRedo() {
        return shapes.getPointer() < shapes.size();
    }
}

File: app/src/main/java/sfllhkhan95/doodle/views/PaintView.kt
package sfllhkhan95.doodle.views

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.Canvas
import android.graphics.PointF
import android.util.AttributeSet
import android.view.MotionEvent
import android.view.View
import sfllhkhan95.doodle.bo.factory.ToolFactory
import sfllhkhan95.doodle.models.*
import sfllhkhan95.doodle.models.shapes.Eraser
import sfllhkhan95.doodle.models.shapes.FloodFill
import sfllhkhan95.doodle.utils.listener.CanvasActionListener
import sfllhkhan95.doodle.utils.listener.OnColorPickedListener


/**
 * @author saifkhichi96
 */
class PaintView : View {

    val brush: PaintBrush
    var canvas: PaintCanvas? = null

    var selectedTool: Class<out Tool>? = null
    private val tools = Tools()

    private val iTouch = PointF()
    private var fTouch: PointF? = PointF()

    private var canvasActionListener: CanvasActionListener? = null
    private var onColorPickedListener: OnColorPickedListener? = null

    val isModified: Boolean
        get() = tools.pointer > 0

    constructor(context: Context) : super(context) {
        brush = PaintBrush(context)
    }

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
        brush = PaintBrush(context)
    }

    override fun onDraw(canvas: Canvas) {
        canvas.save()
        this.canvas?.let {
            it.drawBackground()
            it.drawPaths(tools)

            canvas.drawBitmap(it.bitmap, 0f, 0f, null)
        }
        canvas.restore()
    }

    @SuppressLint("ClickableViewAccessibility")
    override fun onTouchEvent(ev: MotionEvent): Boolean {
        if (selectedTool == null || !isEnabled || ev.pointerCount > 1) return false
        val touchAt = PointF(ev.x, ev.y)

        when (ev.action) {
            MotionEvent.ACTION_DOWN -> {
                touchStart(touchAt)
                invalidate()
                return true
            }
            MotionEvent.ACTION_MOVE -> {
                touchMove(touchAt)
                invalidate()
                return true
            }
            MotionEvent.ACTION_UP -> {
                touchUp(touchAt)
                performClick()
                invalidate()
                return true
            }
        }

        return false
    }

    fun canRedo(): Boolean {
        return tools.pointer < tools.size
    }

    fun clear() {
        tools.clear()
        invalidate()

        canvasActionListener?.onRevert()
    }

    fun redo() {
        val canRedo = tools.redo()
        invalidate()

        canvasActionListener?.onRedo(canRedo)
    }

    fun setCanvasActionListener(canvasActionListener: CanvasActionListener) {
        this.canvasActionListener = canvasActionListener
    }

    fun setOnColorPickedListener(onColorPickedListener: OnColorPickedListener) {
        this.onColorPickedListener = onColorPickedListener
    }

    fun save() {
        canvas?.saveProject()
    }

    fun saveAs() {
        canvas?.saveProjectCopy()
    }

    fun undo() {
        val canUndo = tools.undo()
        invalidate()

        canvasActionListener?.onUndo(canUndo)
    }

    private fun touchStart(touchAt: PointF) {
        if (this.selectedTool == EyedropTool::class.java) {
            canvas?.let {
                val color = it.getColor(touchAt)
                this.onColorPickedListener?.onColorPicked(color)
                return
            }
        }

        iTouch.set(touchAt)
        fTouch = null
    }

    private fun touchMove(touchAt: PointF) {
        if (this.selectedTool == EyedropTool::class.java) {
            canvas?.let {
                val color = it.getColor(touchAt)
                this.onColorPickedListener?.onColorPicked(color)
                return
            }
        }

        if (fTouch == null) {
            fTouch = PointF()

            val tool = ToolFactory[selectedTool!!, brush]
            tool!!.moveTo(iTouch.x, iTouch.y)
            canvas?.let {
                when (tool.javaClass) {
                    Eraser::class.java -> (tool as Eraser).initEraser(it)
                    FloodFill::class.java -> {
                        (tool as FloodFill).apply {
                            this.bitmap = it.bitmap
                            this.execute(iTouch)
                        }
                    }
                }
                tools.add(tool)
            }
        }

        fTouch?.let {
            val dx = Math.abs(touchAt.x - it.x)
            val dy = Math.abs(touchAt.y - it.y)
            if (dx < TOUCH_TOLERANCE && dy < TOUCH_TOLERANCE) {
                return
            }

            it.set(touchAt)
            tools.current?.draw(iTouch, it)
        }
    }

    private fun touchUp(touchAt: PointF) {
        canvas?.let { c ->
            if (this.selectedTool == EyedropTool::class.java) {
                val color = c.getColor(touchAt)
                this.onColorPickedListener?.onColorPicked(color)
                return
            }

            if (fTouch == null) {
                fTouch = touchAt
                val tool = ToolFactory[selectedTool!!, brush]
                tool!!.moveTo(touchAt.x, touchAt.y)
                if (tool.javaClass == Eraser::class.java) {
                    (tool as Eraser).initEraser(c)
                }
                tools.add(tool)
            }

            fTouch?.let {
                val currentTool = tools.current
                currentTool?.draw(iTouch, it)
                canvasActionListener?.onDrawPath()
            }
        }
    }

    companion object {
        private const val TOUCH_TOLERANCE = 5f
    }

}